File: chairlift/core/homebrew.py
"""
Homebrew package manager interface.

This module provides functions to interact with Homebrew package manager,
including listing installed packages, checking for outdated packages,
and searching for available formulae.
"""

import subprocess
import json
import urllib.request
import urllib.error
from typing import List, Dict, Optional, Any


class HomebrewError(Exception):
    """Base exception for Homebrew-related errors."""
    pass


class HomebrewNotFoundError(HomebrewError):
    """Raised when Homebrew is not installed or not found in PATH."""
    pass


def _run_brew_command(args: List[str], timeout: int = 30) -> str:
    """
    Run a brew command and return the output.
    
    Args:
        args: List of command arguments (e.g., ['list', '--formula'])
        timeout: Command timeout in seconds
        
    Returns:
        Command output as string
        
    Raises:
        HomebrewNotFoundError: If brew is not found
        HomebrewError: If command fails
    """
    try:
        result = subprocess.run(
            ['brew'] + args,
            capture_output=True,
            text=True,
            timeout=timeout,
            check=True
        )
        return result.stdout
    except FileNotFoundError:
        raise HomebrewNotFoundError("Homebrew not found. Please install Homebrew first.")
    except subprocess.TimeoutExpired:
        raise HomebrewError(f"Command 'brew {' '.join(args)}' timed out after {timeout} seconds")
    except subprocess.CalledProcessError as e:
        raise HomebrewError(f"Brew command failed: {e.stderr}")


def is_homebrew_installed() -> bool:
    """
    Check if Homebrew is installed and accessible.
    
    Returns:
        True if Homebrew is installed, False otherwise
    """
    try:
        subprocess.run(
            ['brew', '--version'],
            capture_output=True,
            timeout=5,
            check=True
        )
        return True
    except (FileNotFoundError, subprocess.CalledProcessError, subprocess.TimeoutExpired):
        return False


def list_installed_packages(formula_only: bool = True) -> List[Dict[str, Any]]:
    """
    List all installed Homebrew packages.
    
    Args:
        formula_only: If True, only list formulae (default). If False, include casks too.
        
    Returns:
        List of dictionaries containing package information with keys:
        - name: Package name
        - version: Installed version
        - installed_on_request: Whether explicitly installed by user
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    args = ['info', '--installed', '--json=v2']
    if formula_only:
        args.append('--formula')
    
    output = _run_brew_command(args)
    
    try:
        data = json.loads(output)
        packages = []
        
        # The JSON structure has 'formulae' and 'casks' keys
        items = []
        if formula_only:
            items = data.get('formulae', [])
        else:
            items = data.get('formulae', []) + data.get('casks', [])
        
        for item in items:
            # Get the installed version
            installed_versions = item.get('installed', [])
            version = ''
            if installed_versions:
                if isinstance(installed_versions[0], dict):
                    version = installed_versions[0].get('version', '')
                else:
                    version = str(installed_versions[0])
            
            # Ensure name is a string, not a list
            name = item.get('name', '')
            if isinstance(name, list):
                name = name[0] if name else ''
            
            packages.append({
                'name': str(name),
                'version': version,
                'desc': item.get('desc', ''),
                'pinned': item.get('pinned', False),
                'installed_on_request': item.get('installed_on_request', False)
            })
        return packages
    except json.JSONDecodeError as e:
        raise HomebrewError(f"Failed to parse brew output: {e}")


def list_outdated_packages(formula_only: bool = True) -> List[Dict[str, Any]]:
    """
    List outdated Homebrew packages that have updates available.
    
    Args:
        formula_only: If True, only list formulae (default). If False, include casks too.
        
    Returns:
        List of dictionaries containing outdated package information with keys:
        - name: Package name
        - current_version: Currently installed version
        - latest_version: Latest available version
        - pinned: Whether the package is pinned
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    args = ['outdated', '--json=v2']
    if formula_only:
        args.append('--formula')
    
    output = _run_brew_command(args)
    
    try:
        data = json.loads(output)
        outdated = []
        
        # Handle both formulae and casks
        items = data.get('formulae', [])
        if not formula_only:
            items.extend(data.get('casks', []))
        
        for item in items:
            outdated.append({
                'name': item.get('name', ''),
                'current_version': item.get('installed_versions', [''])[0] if item.get('installed_versions') else '',
                'latest_version': item.get('current_version', ''),
                'pinned': item.get('pinned', False)
            })
        return outdated
    except json.JSONDecodeError as e:
        raise HomebrewError(f"Failed to parse brew output: {e}")


def search_formula(query: str, limit: Optional[int] = None) -> List[Dict[str, str]]:
    """
    Search for available Homebrew formulae using the Homebrew API.
    Exact matches are prioritized and returned first.
    
    Args:
        query: Search query string
        limit: Maximum number of results to return (None for all)
        
    Returns:
        List of dictionaries containing search results with keys:
        - name: Formula name
        - description: Formula description
        
    Raises:
        HomebrewError: If the API request fails
    """
    try:
        # Fetch formula data from Homebrew API
        api_url = "https://formulae.brew.sh/api/formula.json"
        
        with urllib.request.urlopen(api_url, timeout=10) as response:
            if response.status != 200:
                raise HomebrewError(f"API request failed with status {response.status}")
            
            data = json.loads(response.read().decode('utf-8'))
        
        # Search through formulae - prioritize exact matches
        exact_matches = []
        partial_matches = []
        query_lower = query.lower()
        
        for formula in data:
            # Get fields
            name = formula.get('name', '')
            full_name = formula.get('full_name', '')
            desc = formula.get('desc', '')
            
            name_lower = name.lower()
            full_name_lower = full_name.lower()
            desc_lower = desc.lower()
            
            # Check for exact match first (name or full_name)
            if name_lower == query_lower or full_name_lower == query_lower:
                exact_matches.append({
                    'name': name,
                    'description': desc
                })
            # Check if query matches any of the fields (partial match)
            elif (query_lower in name_lower or 
                  query_lower in full_name_lower or 
                  query_lower in desc_lower):
                
                partial_matches.append({
                    'name': name,
                    'description': desc
                })
        
        # Combine exact matches first, then partial matches
        results = exact_matches + partial_matches
        
        # Apply limit if specified
        if limit:
            results = results[:limit]
        
        return results
        
    except urllib.error.URLError as e:
        raise HomebrewError(f"Failed to access Homebrew API: {e}")
    except json.JSONDecodeError as e:
        raise HomebrewError(f"Failed to parse API response: {e}")
    except Exception as e:
        raise HomebrewError(f"Search failed: {e}")


def get_package_info(package_name: str) -> Optional[Dict[str, Any]]:
    """
    Get detailed information about a specific package.
    
    Args:
        package_name: Name of the package
        
    Returns:
        Dictionary containing package information with keys:
        - name: Package name
        - version: Latest version
        - description: Package description
        - homepage: Homepage URL
        - installed: Whether currently installed
        - installed_version: Installed version (if installed)
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    output = _run_brew_command(['info', '--json=v2', package_name])
    
    try:
        data = json.loads(output)
        formulae = data.get('formulae', [])
        
        if not formulae:
            return None
        
        formula = formulae[0]
        installed_info = formula.get('installed', [])
        
        return {
            'name': formula.get('name', ''),
            'version': formula.get('versions', {}).get('stable', ''),
            'description': formula.get('desc', ''),
            'homepage': formula.get('homepage', ''),
            'installed': len(installed_info) > 0,
            'installed_version': installed_info[0].get('version', '') if installed_info else None
        }
    except json.JSONDecodeError as e:
        raise HomebrewError(f"Failed to parse brew output: {e}")


def update_homebrew() -> bool:
    """
    Update Homebrew itself and all formulae definitions.
    
    Returns:
        True if update was successful
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    _run_brew_command(['update'], timeout=120)
    return True


def upgrade_package(package_name: Optional[str] = None) -> bool:
    """
    Upgrade a specific package or all outdated packages.
    
    Args:
        package_name: Name of package to upgrade (None to upgrade all)
        
    Returns:
        True if upgrade was successful
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    args = ['upgrade']
    if package_name:
        args.append(package_name)
    
    _run_brew_command(args, timeout=300)
    return True


def pin_package(package_name: str) -> bool:
    """
    Pin a package to prevent it from being upgraded.
    
    Args:
        package_name: Name of package to pin
        
    Returns:
        True if pin was successful
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    _run_brew_command(['pin', package_name])
    return True


def unpin_package(package_name: str) -> bool:
    """
    Unpin a package to allow it to be upgraded.
    
    Args:
        package_name: Name of package to unpin
        
    Returns:
        True if unpin was successful
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    _run_brew_command(['unpin', package_name])
    return True


def uninstall_package(package_name: str, force: bool = False) -> bool:
    """
    Uninstall a package.
    
    Args:
        package_name: Name of package to uninstall
        force: If True, force removal even if other packages depend on it
        
    Returns:
        True if uninstall was successful
        
    Raises:
        HomebrewNotFoundError: If Homebrew is not installed
        HomebrewError: If the command fails
    """
    args = ['uninstall', package_name]
    if force:
        args.append('--force')
    
    _run_brew_command(args, timeout=120)
    return True

File: internal/homebrew/homebrew.go
// Package homebrew provides an interface to the Homebrew package manager
package homebrew

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"log"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"
)

const (
	// readTimeout bounds read-only brew commands (listing, searching, info).
	readTimeout = 30 * time.Second
	// mutationTimeout bounds state-changing brew commands, which may download
	// and build packages and therefore need a far larger budget.
	mutationTimeout = 30 * time.Minute
	// waitDelay bounds how long Wait blocks after the process group has been
	// signalled. brew's helpers (git, curl, download workers) inherit the
	// stdout/stderr pipes, so a straggler could otherwise hold Wait open
	// forever even though the command itself is gone.
	waitDelay = 5 * time.Second
)

var dryRun = false

// SetDryRun sets the dry-run mode
func SetDryRun(mode bool) {
	dryRun = mode
	log.Printf("Homebrew dry-run mode: %v", mode)
}

// IsDryRun returns whether dry-run mode is enabled
func IsDryRun() bool {
	return dryRun
}

// Error represents a Homebrew-related error. Err, when non-nil, carries the
// underlying cause (for example context.DeadlineExceeded or
// context.Canceled) so callers can classify it with errors.Is.
type Error struct {
	Message string
	Err     error
}

func (e *Error) Error() string {
	return e.Message
}

// Unwrap exposes the underlying cause to errors.Is/errors.As.
func (e *Error) Unwrap() error {
	return e.Err
}

// NotFoundError is returned when Homebrew is not installed
type NotFoundError struct {
	Message string
}

func (e *NotFoundError) Error() string {
	return e.Message
}

// Package represents an installed Homebrew package
type Package struct {
	Name               string   `json:"name"`
	Version            string   `json:"version"`
	InstalledOnRequest bool     `json:"installed_on_request"`
	Pinned             bool     `json:"pinned"`
	Outdated           bool     `json:"outdated"`
	Dependencies       []string `json:"dependencies,omitempty"`
}

// SearchResult represents a search result
type SearchResult struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Homepage    string `json:"homepage"`
}

// stateChangingCommands are commands that modify system state
var stateChangingCommands = map[string]bool{
	"install":   true,
	"uninstall": true,
	"remove":    true,
	"upgrade":   true,
	"update":    true,
	"pin":       true,
	"unpin":     true,
	"bundle":    true,
	"cleanup":   true,
	"trust":     true,
}

// commandTimeout returns the timeout class for a brew invocation: the
// mutation timeout for state-changing commands, the read timeout otherwise.
func commandTimeout(args []string) time.Duration {
	if len(args) > 0 && stateChangingCommands[args[0]] {
		return mutationTimeout
	}
	return readTimeout
}

// runBrewCommand executes a brew command and returns the output
func runBrewCommand(args ...string) (string, error) {
	if len(args) > 0 && stateChangingCommands[args[0]] && dryRun {
		msg := fmt.Sprintf("[DRY-RUN] Would execute: brew %s", strings.Join(args, " "))
		log.Println(msg)
		return msg, nil
	}

	ctx, cancel := context.WithTimeout(context.Background(), commandTimeout(args))
	defer cancel()

	return runBrewCommandAt(ctx, "brew", args...)
}

// runBrewCommandAt runs exe with args under ctx and returns its stdout. The
// executable and context are parameters so tests can drive a fake script and
// control the deadline; runBrewCommand is the only production caller and
// always passes "brew".
//
// The command runs in its own process group and cancellation signals the
// whole group, so brew's helper processes (git, curl, download workers) die
// with it rather than being orphaned. cmd.Run still reaps the child.
func runBrewCommandAt(ctx context.Context, exe string, args ...string) (string, error) {
	cmd := exec.CommandContext(ctx, exe, args...)
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	cmd.Cancel = func() error {
		return syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
	}
	cmd.WaitDelay = waitDelay

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		display := exe
		if len(args) > 0 {
			display += " " + strings.Join(args, " ")
		}
		// Classify the context outcome first: the process was killed by our
		// own Cancel func, so the raw error would otherwise read
		// "signal: killed".
		if errors.Is(ctx.Err(), context.DeadlineExceeded) || errors.Is(err, context.DeadlineExceeded) {
			return "", &Error{
				Message: fmt.Sprintf("Command '%s' timed out", display),
				Err:     context.DeadlineExceeded,
			}
		}
		if errors.Is(ctx.Err(), context.Canceled) || errors.Is(err, context.Canceled) {
			return "", &Error{
				Message: fmt.Sprintf("Command '%s' was canceled", display),
				Err:     context.Canceled,
			}
		}
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			if isUntrustedTapMessage(stderr.String()) {
				return "", &UntrustedTapError{Message: fmt.Sprintf("Brew command failed: %s", stderr.String())}
			}
			return "", &Error{Message: fmt.Sprintf("Brew command failed: %s", stderr.String()), Err: err}
		}
		// exec.ErrNotFound covers a bare name missing from $PATH;
		// fs.ErrNotExist covers an explicit path that does not exist.
		if errors.Is(err, exec.ErrNotFound) || errors.Is(err, fs.ErrNotExist) {
			return "", &NotFoundError{Message: "Homebrew not found. Please install Homebrew first."}
		}
		return "", &Error{Message: err.Error(), Err: err}
	}

	return stdout.String(), nil
}

// IsInstalled checks if Homebrew is installed and accessible
func IsInstalled() bool {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "brew", "--version")
	err := cmd.Run()
	return err == nil
}

var (
	installedOnce   sync.Once
	installedResult bool
)

// IsInstalledCached returns a cached result of IsInstalled, running the check at most once.
func IsInstalledCached() bool {
	installedOnce.Do(func() {
		installedResult = IsInstalled()
	})
	return installedResult
}

// ListInstalledFormulae returns all installed formulae
func ListInstalledFormulae() ([]Package, error) {
	output, err := runBrewCommand("info", "--installed", "--json=v2", "--formula")
	if err != nil {
		return nil, err
	}

	return parsePackagesJSON(output, true)
}

// ListInstalledCasks returns all installed casks
func ListInstalledCasks() ([]Package, error) {
	output, err := runBrewCommand("info", "--installed", "--json=v2", "--cask")
	if err != nil {
		return nil, err
	}

	return parsePackagesJSON(output, false)
}

// parsePackagesJSON parses the JSON output from brew info
func parsePackagesJSON(jsonData string, isFormula bool) ([]Package, error) {
	var data struct {
		Formulae []struct {
			Name     string `json:"name"`
			Versions struct {
				Stable string `json:"stable"`
			} `json:"versions"`
			Installed []struct {
				Version            string `json:"version"`
				InstalledOnRequest bool   `json:"installed_on_request"`
			} `json:"installed"`
			Pinned   bool `json:"pinned"`
			Outdated bool `json:"outdated"`
		} `json:"formulae"`
		Casks []struct {
			Token     string `json:"token"`
			Version   string `json:"version"`
			Installed string `json:"installed"`
			Outdated  bool   `json:"outdated"`
		} `json:"casks"`
	}

	if err := json.Unmarshal([]byte(jsonData), &data); err != nil {
		return nil, &Error{Message: fmt.Sprintf("Failed to parse JSON: %v", err)}
	}

	var packages []Package

	if isFormula {
		for _, f := range data.Formulae {
			if len(f.Installed) == 0 {
				continue
			}
			packages = append(packages, Package{
				Name:               f.Name,
				Version:            f.Installed[0].Version,
				InstalledOnRequest: f.Installed[0].InstalledOnRequest,
				Pinned:             f.Pinned,
				Outdated:           f.Outdated,
			})
		}
	} else {
		for _, c := range data.Casks {
			packages = append(packages, Package{
				Name:     c.Token,
				Version:  c.Installed,
				Outdated: c.Outdated,
			})
		}
	}

	return packages, nil
}

// ListOutdated returns all outdated packages
func ListOutdated() ([]Package, error) {
	output, err := runBrewCommand("outdated", "--json=v2")
	if err != nil {
		return nil, err
	}

	var data struct {
		Formulae []struct {
			Name              string   `json:"name"`
			InstalledVersions []string `json:"installed_versions"`
			CurrentVersion    string   `json:"current_version"`
			Pinned            bool     `json:"pinned"`
		} `json:"formulae"`
		Casks []struct {
			Name              string   `json:"name"`
			InstalledVersions []string `json:"installed_versions"`
			CurrentVersion    string   `json:"current_version"`
		} `json:"casks"`
	}

	if err := json.Unmarshal([]byte(output), &data); err != nil {
		return nil, &Error{Message: fmt.Sprintf("Failed to parse JSON: %v", err)}
	}

	var packages []Package
	for _, f := range data.Formulae {
		packages = append(packages, Package{
			Name:     f.Name,
			Version:  strings.Join(f.InstalledVersions, ", "),
			Outdated: true,
			Pinned:   f.Pinned,
		})
	}
	for _, c := range data.Casks {
		packages = append(packages, Package{
			Name:     c.Name,
			Version:  strings.Join(c.InstalledVersions, ", "),
			Outdated: true,
		})
	}

	return packages, nil
}

// Search searches for formulae matching the query
func Search(query string) ([]SearchResult, error) {
	output, err := runBrewCommand("search", "--formula", query)
	if err != nil {
		return nil, err
	}

	var results []SearchResult
	lines := strings.Split(strings.TrimSpace(output), "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line != "" && !strings.HasPrefix(line, "==>") {
			results = append(results, SearchResult{Name: line})
		}
	}

	return results, nil
}

// Install installs a package
func Install(name string, isCask bool) error {
	args := []string{"install"}
	if isCask {
		args = append(args, "--cask")
	}
	args = append(args, name)

	_, err := runBrewCommand(args...)
	return err
}

// Uninstall removes a package
func Uninstall(name string, isCask bool) error {
	args := []string{"uninstall"}
	if isCask {
		args = append(args, "--cask")
	}
	args = append(args, name)

	_, err := runBrewCommand(args...)
	return err
}

// Upgrade upgrades a package or all packages
func Upgrade(name string) error {
	args := []string{"upgrade"}
	if name != "" {
		args = append(args, name)
	}

	_, err := runBrewCommand(args...)
	return err
}

// Update updates Homebrew itself
func Update() error {
	_, err := runBrewCommand("update")
	return err
}

// Pin pins a package
func Pin(name string) error {
	_, err := runBrewCommand("pin", name)
	return err
}

// Unpin unpins a package
func Unpin(name string) error {
	_, err := runBrewCommand("unpin", name)
	return err
}

// BundleDump dumps installed packages to a Brewfile
func BundleDump(path string, force bool) error {
	args := []string{"bundle", "dump"}
	if path != "" {
		args = append(args, "--file="+path)
	}
	if force {
		args = append(args, "--force")
	}

	_, err := runBrewCommand(args...)
	return err
}

// BundleInstall installs packages from a Brewfile
func BundleInstall(path string) error {
	args := []string{"bundle", "install"}
	if path != "" {
		args = append(args, "--file="+path)
	}

	_, err := runBrewCommand(args...)
	return err
}

// Cleanup removes old versions, outdated downloads, and clears cache
func Cleanup() (string, error) {
	return runBrewCommand("cleanup")
}
